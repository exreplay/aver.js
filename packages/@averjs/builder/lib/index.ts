import StaticBuilder from './builders/static';
import SsrBuilder from './builders/ssr';
import { BuilderContext } from './builders/base';

export { StaticBuilder, SsrBuilder, BuilderContext };
