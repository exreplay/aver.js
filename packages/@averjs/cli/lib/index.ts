import HelpCommand from './commands/help';
import DevCommand from './commands/dev';
import InitCommand from './commands/init';
import BuildCommand from './commands/build';
import ProductionCommand from './commands/prod';
import parseArgs, { ParsedArgs } from 'minimist';
import {
  Args,
  CommandInterface,
  CommandInterfaceDictionary
} from './commands/command';
import './global';

export { CommandInterface };

export default class Usage {
  argv: ParsedArgs;
  aliases: { [arg: string]: string } = {};
  availableCommands: CommandInterfaceDictionary = {};

  get help() {
    return this.argv._[0] === 'help' || this.argv.help;
  }

  get executedCommand() {
    return this.argv._[0] || (!this.help ? 'dev' : 'help');
  }

  get globalCommand(): Args | undefined {
    return this.availableCommands.help.args?.find(
      option => this.argv[option.name]
    );
  }

  constructor() {
    this.addCommand(new HelpCommand(this.availableCommands));
    this.addCommand(new DevCommand());
    this.addCommand(new ProductionCommand());
    this.addCommand(new InitCommand());
    this.addCommand(new BuildCommand());

    this.argv = parseArgs(process.argv.slice(2), {
      alias: this.aliases
    });
  }

  addCommand(command: CommandInterface) {
    this.availableCommands[command.name] = command;

    for (const alias of command.aliases || []) {
      this.aliases[alias] = command.name;

      // Register aliases for command options
      for (const argAlias of command.args || []) {
        if (argAlias.alias) this.aliases[argAlias.alias] = argAlias.name;
      }
    }
  }

  async run() {
    const commandToExecute = this.availableCommands[this.executedCommand];

    // If a global option is found, execute it and stop afterwards
    if (this.globalCommand?.command) {
      await this.globalCommand.command.run();
      return;
    }

    try {
      // No matter how help is set, do not run the actual command, instead show default help or for specified command
      if (this.help)
        await this.availableCommands.help.run(
          this.executedCommand !== 'help' ? commandToExecute : undefined
        );
      else await commandToExecute.run(this.argv);
    } catch (error) {
      console.error(error);
    }
  }
}
