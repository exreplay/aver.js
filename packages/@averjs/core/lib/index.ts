import Core from './core';
import path from 'path';
export * from './plugins';

/**
 * Remove in next major version, where a index.js file in root directory is not needed anymore.
 * Those variables are defined in the executable in the cli package.
 */
process.env.PROJECT_PATH = path.resolve(process.cwd(), './src');
process.env.API_PATH = path.resolve(process.cwd(), './api');

export default Core;
