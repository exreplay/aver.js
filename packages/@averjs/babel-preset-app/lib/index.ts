// Props to vue-cli
// TODO: Add modern mode
import { PluginObj } from '@babel/core';
import PresetEnv, { Options as BabelPresetOptions } from '@babel/preset-env';
import Polyfills from './polyfills';
import PolyfillsPlugin from './polyfillsPlugin';
import PluginSyntaxDynamicImport from '@babel/plugin-syntax-dynamic-import';
import PluginProposalDecorators from '@babel/plugin-proposal-decorators';
import PluginProposalClassProperties from '@babel/plugin-proposal-class-properties';
import PluginTransformClasses from '@babel/plugin-transform-classes';
import PluginTransformRuntime from '@babel/plugin-transform-runtime';

export type BabelPlugin =
  | [NodeRequire | string | (() => PluginObj), Record<string, unknown>]
  | (NodeRequire | string | (() => PluginObj));

export type BabelOptions = BabelPresetOptions & {
  polyfills?: string[];
  buildTarget?: 'client' | 'server';
  decoratorsBeforeExport?: boolean;
  decoratorsLegacy?: boolean;
  absoluteRuntime?: boolean | string;
};

class BabelPresetApp extends Polyfills {
  options: BabelOptions;

  get targets() {
    return this.options.buildTarget === 'server'
      ? { node: 'current' }
      : {
          ie: '9'
        };
  }

  get corejs() {
    return this.options.corejs && typeof this.options.corejs !== 'number'
      ? this.options.corejs?.version
      : this.options.corejs || 2;
  }

  get loose() {
    return this.options.loose || true;
  }

  get debug() {
    return this.options.debug || false;
  }

  get useBuiltIns() {
    return this.options.useBuiltIns || 'usage';
  }

  get modules() {
    return this.options.modules || false;
  }

  get ignoreBrowserslistConfig() {
    return this.options.ignoreBrowserslistConfig || false;
  }

  get decoratorsLegacy() {
    return this.options.decoratorsLegacy;
  }

  constructor(options: BabelOptions) {
    super();
    this.options = options;
  }

  presets(polyfills: Required<BabelOptions>['polyfills']) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const presets: any[] = [];

    presets.push([
      PresetEnv,
      {
        spec: this.options.spec,
        loose: this.loose,
        debug: this.debug,
        modules: this.modules,
        targets: this.targets,
        useBuiltIns: this.useBuiltIns,
        corejs: this.corejs,
        ignoreBrowserslistConfig: this.ignoreBrowserslistConfig,
        configPath: this.options.configPath,
        include: this.options.include,
        exclude: polyfills.concat(
          this.options.exclude?.filter(
            /* istanbul ignore next */ (e): e is string => typeof e === 'string'
          ) || []
        ),
        shippedProposals: this.options.shippedProposals,
        forceAllTransforms: this.options.forceAllTransforms
      } as BabelPresetOptions
    ]);

    return presets;
  }

  plugins() {
    const plugins: BabelPlugin[] = [];

    plugins.push(PluginSyntaxDynamicImport);

    plugins.push([
      PluginProposalDecorators,
      {
        decoratorsBeforeExport: this.options.decoratorsBeforeExport,
        legacy: this.decoratorsLegacy !== false
      }
    ]);

    plugins.push([PluginProposalClassProperties, { loose: this.loose }]);
    plugins.push([PluginTransformClasses, { loose: this.loose }]);

    plugins.push([
      PluginTransformRuntime,
      {
        regenerator: this.useBuiltIns !== 'usage',
        // polyfills are injected by preset-env & polyfillsPlugin, so no need to add them again
        corejs: false,
        helpers: this.useBuiltIns === 'usage',
        useESModules: this.options.buildTarget !== 'server',
        absoluteRuntime: this.options.absoluteRuntime
      }
    ]);

    return plugins;
  }

  polyfills(plugins: BabelPlugin[]) {
    let polyfills: BabelOptions['polyfills'] = [];

    if (this.useBuiltIns === 'usage' && this.options.buildTarget === 'client') {
      polyfills = this.getPolyfills(
        this.corejs,
        this.targets,
        this.options.polyfills || this.getDefaultPolyfills(this.corejs),
        {
          ignoreBrowserslistConfig: this.ignoreBrowserslistConfig,
          configPath: this.options.configPath
        }
      );
      plugins.push([PolyfillsPlugin, { polyfills }]);
    } else {
      polyfills = [];
    }

    return polyfills;
  }

  config() {
    const plugins = this.plugins();
    const polyfills = this.polyfills(plugins);
    const presets = this.presets(polyfills);

    return {
      sourceType: 'unambiguous',
      presets,
      plugins
    };
  }
}

export default (_api: never, options: BabelOptions = {}) =>
  new BabelPresetApp(options).config();
