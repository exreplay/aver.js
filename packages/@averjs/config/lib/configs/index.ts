import app from './app';
import core from './core';
import server from './server';
import renderer from './renderer';
import vueApp from './vue-app';

export const defaultFileName = 'aver-config';

type AverConfig = ReturnType<typeof app> & ReturnType<typeof core> & ReturnType<typeof server> & ReturnType<typeof vueApp> & {
  webpack: ReturnType<typeof renderer>
}

export function defaultAverjsConfig(isProd: boolean): AverConfig {
  return {
    ...app(),
    ...core(),
    ...server(),
    webpack: renderer(isProd),
    ...vueApp()
  };
}
