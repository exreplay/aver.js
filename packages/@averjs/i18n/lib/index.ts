import path from 'path';
import { PluginFunction } from '@averjs/core';

const plugin: PluginFunction = function () {
  this.config.templates?.push({
    src: path.resolve(__dirname, '../entries/i18n.js'),
    dst: path.resolve(this.config.cacheDir, './i18n/i18n.js')
  });

  this.config.templates?.push({
    src: path.resolve(__dirname, '../entries/i18n.d.ts'),
    dst: path.resolve(this.config.cacheDir, './i18n/i18n.d.ts')
  });

  this.config.templates?.push({
    src: path.resolve(__dirname, '../entries/app.js'),
    dst: path.resolve(this.config.cacheDir, './i18n/app.js')
  });

  this.config.i18n = {
    silentTranslationWarn: true,
    ...this.config.i18n
  };

  if (!this.config.additionalTemplatesConfig)
    this.config.additionalTemplatesConfig = {};

  this.config.additionalTemplatesConfig.i18n = this.config.i18n;

  if (this.aver.config.webpack?.alias)
    this.aver.config.webpack.alias['@i18n'] = path.resolve(
      this.config.cacheDir,
      './i18n/i18n.js'
    );

  this.aver.tap('renderer:base-config', (chain) => {
    chain.module
      .rule('i18n')
      .resourceQuery(/blockType=i18n/)
      .type('javascript/auto')
      .use('i18n')
      .loader('@kazupon/vue-i18n-loader');
  });
};

export default plugin;
