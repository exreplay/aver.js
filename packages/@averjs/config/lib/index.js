import path from 'path';
import mergeWith from 'lodash/mergeWith';
import { defaultAverjsConfig, defaultFileName } from './configs';

export function getAverjsConfig() {
  const requireModule = require('esm')(module);
  const globalConfPath = path.resolve(process.env.PROJECT_PATH, `../${defaultFileName}`);
  const config = defaultAverjsConfig();
  let userConfig = {};
  let configFile = null;

  try {
    configFile = require.resolve(globalConfPath);
  } catch (err) {
    if (err.code !== 'MODULE_NOT_FOUND') throw err;
    else console.log('Could not find aver-config file. Proceeding with default config...');
  }

  if (configFile) {
    if (process.env.NODE_ENV === 'test') userConfig = require(configFile).default;
    else userConfig = requireModule(configFile).default;
  }

  config.rootDir = process.cwd();

  config.cacheDir = path.resolve(config.rootDir, './node_modules/.cache/averjs');

  config.distDir = './dist';
  config.distPath = path.resolve(config.rootDir, config.distDir);

  return mergeWith(config, userConfig, (objValue, srcValue) => {
    if (Array.isArray(objValue)) {
      return objValue.concat(srcValue);
    }
  });
}
