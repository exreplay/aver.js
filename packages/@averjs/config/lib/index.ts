/* eslint-disable @typescript-eslint/no-var-requires */
import path from 'path';
import mergeWith from 'lodash/mergeWith';
import { defaultAverjsConfig, defaultFileName } from './configs';

interface InternalConfig {
  rootDir: string;
  cacheDir: string;
  distPath: string;
  distDir: string;
  isProd: boolean;
  _production: boolean;
}

export type AverConfig = ReturnType<typeof defaultAverjsConfig> &
  InternalConfig & { [index: string]: any };

export function getAverjsConfig() {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const requireModule = require('esm')(module);
  const globalConfPath = path.resolve(
    process.env.PROJECT_PATH,
    `../${defaultFileName}`
  );
  const isProd =
    process.env.NODE_ENV === 'production' || process.env.NODE_ENV === 'test';
  const config = defaultAverjsConfig(isProd) as AverConfig;
  let userConfig = {};
  let configFile = null;

  try {
    configFile = require.resolve(globalConfPath);
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code !== 'MODULE_NOT_FOUND')
      throw error;
    else
      console.log(
        'Could not find aver-config file. Proceeding with default config...'
      );
  }

  if (configFile) {
    if (process.env.NODE_ENV === 'test')
      userConfig = require(configFile).default;
    /* istanbul ignore next */ else
      userConfig = requireModule(configFile).default;
  }

  config.isProd = isProd;

  config.rootDir = process.cwd();

  config.cacheDir = path.resolve(
    config.rootDir,
    './node_modules/.cache/averjs'
  );

  config.distDir = './dist';
  config.distPath = path.resolve(config.rootDir, config.distDir);

  return mergeWith(config, userConfig, (objValue, srcValue) => {
    if (Array.isArray(objValue)) {
      return objValue.concat(srcValue);
    }
  });
}
