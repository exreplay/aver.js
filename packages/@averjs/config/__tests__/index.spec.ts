import path from 'path';
import { getAverjsConfig } from '../lib/index';

beforeEach(() => {
  console.log = jest.fn();
});

beforeEach(() => {
  process.env.PROJECT_PATH = path.resolve(__dirname, '../__fixtures__/src');
});

test('should change progressbar option to false', () => {
  const config = getAverjsConfig();
  expect(config.progressbar).toBe(false);
  expect(config.webpack.css?.extract).toBe(true);
});

test('should concat array with default values', () => {
  const config = getAverjsConfig();
  expect(config.progressbar).toBe(false);
  expect(config.webpack.additionalExtensions).toEqual(['js', 'ts']);
});

test('should match default snapshot config when no config file is present', () => {
  const processSpy = jest.spyOn(process, 'cwd').mockReturnValue('/');
  
  process.env.PROJECT_PATH = '/aver';
  const config = getAverjsConfig();
  expect(config).toMatchSnapshot();

  processSpy.mockRestore();
});
