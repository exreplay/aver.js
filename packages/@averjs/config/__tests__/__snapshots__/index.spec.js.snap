// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match default snapshot config when no config file is present 1`] = `
Object {
  "aliases": Object {},
  "buildPlugins": Array [],
  "cacheDir": "/node_modules/.cache/averjs",
  "createRenderer": Object {},
  "csrf": true,
  "csrfExclude": Array [],
  "distDir": "./dist",
  "distPath": "/dist",
  "entries": Object {
    "app": Array [],
    "client": Array [],
    "server": Array [],
  },
  "i18n": Object {
    "silentTranslationWarn": true,
  },
  "plugins": Array [],
  "progressbar": true,
  "rootDir": "/",
  "store": Object {},
  "templates": Array [],
  "webpack": Object {
    "additionalExtensions": Array [
      "js",
    ],
    "alias": Object {
      "@": "/aver",
      "@@": "/",
      "@components": "/aver/components",
      "@mixins": "/aver/mixins",
      "@pages": "/aver/pages",
      "@resources": "/aver/resources",
      "@vuex": "/aver/vuex",
    },
    "babel": Object {},
    "base": false,
    "client": false,
    "css": Object {
      "extract": false,
      "styleResources": Object {
        "options": Object {},
        "resources": Array [],
      },
    },
    "postcss": Object {},
    "process": Object {
      "env": Object {},
    },
    "server": false,
    "sw": false,
    "transpileDependencies": Array [],
  },
}
`;
